docs/provider-pkcs11.7.md

```diff
@@ -0,0 +1,262 @@
+% provider-pkcs11(7) | Configuration directives
+
+NAME
+====
+pkcs11-provider - An OpenSSL provider that allows to directly interface
+with pkcs11 drivers.
+
+
+DESCRIPTION
+===========
+
+Starting with version 3.0 the OpenSSL project introduced a new modular
+system to exten OpenSSL that replaces the deprecated Engines modules
+
+Providers[1] are loaded via configuration directives in the openssl
+configuration file (or via direct loading by applications).
+
+The pkcs11 provider allows openssl linked applications to use keys and
+cryptographic operations from a hardware or software token via their
+PKCS#11[2] driver and the use ok pkcs11 URIs[3].
+
+The pkcs11 provider can be configured to be automatically loaded via
+openssl.cnf
+
+
+CONFIGURATION
+=============
+Configuration options recognized by the provider
+
+## pkcs11-module-path
+
+A file path to the pkcs11 driver to be used
+
+Default: None (but see the environment variables section)
+
+Example:
+
+```pkcs11-module-path = /usr/lib64/opensc-pkcs11.so```
+
+## pkcs11-module-init-args
+
+Non-standard initialization arguments some pkcs11 driver may need.
+Generally not used, but some software tokens like NSS's softokn require
+this.
+
+Default: None
+
+Example:
+
+```pkcs11-module-init-args = configDir=/etc/pki/token```
+
+## pkcs11-module-token-pin
+
+The user PIN to be used with the token.
+If a PIN is not set in configuration it can be asked interactively (if
+the application uses prompters), or it can be specified together with
+the key identifiers in the pkcs11 URI directly.
+When a file is specified the file must be a text file containing just
+the PIN on the first line and a \n terminator.
+
+Default: None
+
+Example:
+
+```pkcs11-module-token-pin = file:/etc/pki/pin.txt```
+```
+cat /etc/pki/pin.txt
+123456
+```
+
+## pkcs11-module-allow-export
+
+Whether the pkcs11 provider will allow to export public keys through
+OpenSSL.
+OpenSSL often tries to export public keys from non-default providers to
+the default provider, and then use OpenSSL own functions to handle
+whatever operation is associated with the public key.
+This option can be useful to force public key operations to be executed
+on the token, for example in case the pkcs11 is an accelerator that has
+better performance on signature checking or asymmetric encryption than
+OpenSSL's code.
+
+Default: 0 (Allow Export)
+
+Example:
+
+```pkcs11-module-allow-export = 1```
+(This disallows export of public keys)
```

Comment:
default allows export and setting 1 disallows? I find this quite confusing ...

Reply:
It is a bit, I might change the code to use a keyword in preference in the future ...
I should have called it "disallow" ...